### ISSUE TITLE
Expose skiko's `renderApi` to users

### ISSUE THREAD

**Issue by SebastianAigner** [JetBrains/compose-multiplatform#382]
Currently `org.jetbrains.skiko.SkikoProperties.renderApi` is actually `internal` and not visible to CFD users.

With the introduction of software rendering, users might want to detect which rendering they are running on. Currently, the fallback renderer only shows a message on the command line:

```
Failed to create Skia OpenGL context!
Falling back to software rendering...
```

The property should probably be exposed. 
cc @Rsedaikin

**Comment by olonho** [comment]
Why this property could be needed?

**Comment by jimgoog** [comment]
Presumably apps might want to disable animations or other performance-intensive graphics operations if the mode is not hardware accelerated?  Or perhaps apps want to show a warning banner to the user to indicate the app is in software mode?  Or if the app wants to include that information in a bug report or to track usage/configuration statistics?  It does seem like a potentially useful piece of information to be able to query.

**Comment by olonho** [comment]
Shouldn't we provide API of such a kind on the level of Compose common or Desktop-only instead? Skiko is currently mostly an implementation detail.

**Comment by jimgoog** [comment]
In my mind, this feels like a low-level detail of the underlying layer (skiko) and not something that we would necessarily expose explicitly as part of a Compose API, although I'd be open to counterarguments.  But regardless, wouldn't we need to make this public in skiko as a prerequisite in order for Common or Desktop to be able to read the value and thus expose it?  

As a general rule of thumb, each layer of Compose should only depend on public APIs of the underlying layer, such that any individual layer could in theory be reimplemented by the user.  We also let the user access the underlying layer whenever they need to dive into the details, much like we suggest everyone use common code, but let them call into android/desktop as needed.  There are going to be plenty of use cases where users need to drop down into skiko/skia/skija code (eg. right now `NativePaint` is a `org.jetbrains.skija.Paint`, presumably on non-jvm it would be a `skiko/Paint`) and that seems perfectly ok.  If someone needs access to `renderApi`, is there a downside to telling them to just grab it from skiko?

Does that make sense, or am I missing something in the intent of the layering here?  If the later, maybe you can elaborate on why and what/where/how in Compose/Desktop you might provide such an API?  Certainly I think we would not want to introduce an ambient for something like this.

**Comment by graycatdeveloper** [comment]
[Updated](https://blog.jetbrains.com/kotlin/2021/06/compose-for-desktop-milestone-4-released/) `System.setProperty("skiko.renderApi", "OPENGL")//SOFTWARE, OPENGL, or METAL`

**Comment by Riekr** [comment]
Dont' know if related (came here from a web search) but when starting jetbrains-toolbox via x2go I today starting getting this error:

```
$ jetbrains-toolbox 
Failed to create Skia OpenGL context!
Cannot init graphic context
```

and I'm forced to use a `kill -9 <pid>` to get back to terminal.
My env is:

```
$ env | grep  GL
LIBGL_ALWAYS_INDIRECT=1
QT_XCB_FORCE_SOFTWARE_OPENGL=1
LIBGL_ALWAYS_SOFTWARE=1
```

Unsetting there env variables lets me start toolbox app and I only get this output in console:
```
$ jetbrains-toolbox 
Cannot create Linux GL context
```

hope it helps.

**Comment by akurasov** [comment]
@Riekr Please report it to Toolbox

@Rsedaikin - FYI

**Comment by linaori** [comment]
@Reikr have you reported this issue? I ran into the exact same error:
```
Toolbox 1.27.0.13477
Failed to create Skia OpenGL context!
Cannot init graphic context
```

**Comment by Riekr** [comment]
@linaori no, I didn't report anything as those are env variables I setted to try increase performance via x2go/nomachine.
On local desktop everything works perfectly (but also via x2go wihout them).

**Comment by StefanOltmann** [comment]
How can we now request the actual used `renderApi` if we don't expose `org.jetbrains.skiko.SkikoProperties.renderApi`? 
What's the alternative?

I also want to check if I have hardware acceleration to decide upon the usage of expensive animations.

For Compose Multiplatform to be neutral to implementation details we could introduce a `isHardwareAccelerated()` - I don't care if it's METAL or OPENGL.

**Comment by StefanOltmann** [comment]
I now have actual users with a Intel HD 4600, which has no DirectX 12 support, and want to show them a warning about the decreased performance.

**Comment by StefanOltmann** [comment]
@SebastianAigner 

This seems to work for me:
```
    singleWindowApplication {
        println("Render API: " + this.window.renderApi)
    }
```

**Comment by igordmn** [comment]
> println("Render API: " + this.window.renderApi)

Yes, it was implemented in Compose 1.0 this way.

**Comment by okushnikov** [comment]
Please check the <a href="https://youtrack.jetbrains.com/issue/CMP-382">following ticket</a> on YouTrack for follow-ups to this issue. GitHub issues will be closed in the coming weeks.
